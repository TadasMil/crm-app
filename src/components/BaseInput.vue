<script setup name="BaseInput" lang="ts">
import { useField } from 'vee-validate'
import { computed, toRef, watch } from 'vue'

const props = defineProps<{
  label: string
  modelValue: string
  name: string
  rules: string
}>()

const emit = defineEmits(['update:modelValue'])

const dynamicRules = computed(() => props.rules)

const name = toRef(props, 'name')

const { errorMessage, value, handleChange } = useField(name, dynamicRules, {
  initialValue: props.modelValue,
  valueProp: 'modelValue',
})

watch(
  () => value.value,
  (newValue) => {
    emit('update:modelValue', newValue)
  },
  { deep: true }
)
</script>

<template>
  <div>
    <label
      class="mb-2 block text-sm font-medium text-primary text-gray-900 dark:text-gray-300"
    >
      {{ label }}
    </label>

    <input
      type="text"
      class="block w-full rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 focus:border-primary"
      :name="name"
      :value="value"
      @input="handleChange"
    />
    <span class="text-sm text-danger">{{ errorMessage }}</span>
  </div>
</template>
