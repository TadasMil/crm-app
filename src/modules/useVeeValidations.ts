import { defineRule } from 'vee-validate'

defineRule('required', (value: string) => {
  if (!value || !value.length) {
    return 'This field is required.'
  }

  return true
})

defineRule('email', (value: string) => {
  if (!value || !value.length) {
    return true
  }

  if (!/^\S+@\S+\.\S+$/.test(value)) {
    return 'This field must be a valid email.'
  }

  return true
})

defineRule('min', (value: string, params: string[]) => {
  if (!value || !value.length) {
    return true
  }

  if (value.length < 3) {
    return 'This field must be at least 3 characters long.'
  }

  return true
})

defineRule('zipcode', (value: string) => {
  if (!value || !value.length) {
    return true
  }

  if (!/((?:LT)[\-])?(\d{5})/.test(value)) {
    return 'This field must be a valid zip code.'
  }

  return true
})
