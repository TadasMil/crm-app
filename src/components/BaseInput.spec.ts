import { render, fireEvent } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'
import userEvent from '@testing-library/user-event'
import '../modules/useVeeValidations.js'

import BaseInput from './BaseInput.vue'

describe('BaseInput.spec.js', () => {
  it('should render correct component', () => {
    const { html } = render(BaseInput, {
      props: {
        label: 'First Name',
        modelValue: '',
        name: 'name',
        rules: 'required',
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should emit movelValue update for input', async () => {
    const { getByRole, emitted } = render(BaseInput, {
      props: {
        label: 'First Name',
        modelValue: '',
        name: 'name',
        rules: 'required',
      },
    })

    const input = getByRole('textbox')
    const user = userEvent.setup()

    await user.type(input, 'my name')

    const lastEmittedValue = emitted()['update:modelValue'].pop()

    expect(emitted()['update:modelValue']).toBeTruthy()
    expect(lastEmittedValue).toStrictEqual(['my name'])
  })

  it('should display required error message given the field is left empty', async () => {
    const { html, getByRole } = render(BaseInput, {
      props: {
        label: 'First Name',
        modelValue: 'My name',
        name: 'name',
        rules: 'required',
      },
    })

    const input = getByRole('textbox')
    const user = userEvent.setup()

    await user.clear(input)

    expect(html()).toContain('This field is required.')
  })

  it('should display email error message given the email with incorrect format', async () => {
    const { html, getByRole } = render(BaseInput, {
      props: {
        label: 'Email',
        modelValue: '',
        name: 'email',
        rules: 'email',
      },
    })

    const input = getByRole('textbox')
    const user = userEvent.setup()

    await user.type(input, 'email')
    await fireEvent.blur(input)

    expect(html()).toContain('This field must be a valid email.')
  })

  it('should display minimum length error message given the input with not enough characters', async () => {
    const { html, getByRole } = render(BaseInput, {
      props: {
        label: 'First Name',
        modelValue: '',
        name: 'name',
        rules: 'min',
      },
    })

    const input = getByRole('textbox')
    const user = userEvent.setup()

    await user.type(input, 'na')
    await fireEvent.blur(input)

    expect(html()).toContain('This field must be at least 3 characters long.')
  })

  it('should display zipcode field error message given the zipcode is not in Lithuanian format', async () => {
    const { html, getByRole } = render(BaseInput, {
      props: {
        label: 'Zip Code',
        modelValue: '',
        name: 'zipcode',
        rules: 'zipcode',
      },
    })

    const input = getByRole('textbox')
    const user = userEvent.setup()

    await user.type(input, 'na')
    await fireEvent.blur(input)

    expect(html()).toContain('This field must be a valid zip code.')
  })
})
