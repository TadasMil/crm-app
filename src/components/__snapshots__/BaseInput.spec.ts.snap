// Vitest Snapshot v1

exports[`BaseInput.spec.js > should render correct component 1`] = `"<div><label class=\\"mb-2 block text-sm font-medium text-primary text-gray-900 dark:text-gray-300\\">First Name</label><input type=\\"text\\" class=\\"block w-full rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 focus:border-primary\\" name=\\"name\\"><span class=\\"text-sm text-danger\\"></span></div>"`;
